models/demos/tg/llama3_70b/tt/llama_attention_galaxy.py

```diff
@@ -467,3 +491,194 @@ def attn_selfout(
+        assert xs.shape[2] % 128 == 0 and xs.shape[2] > 0, "Seqlen must be divisible by 128"
+        _, _, seq_len, _ = xs.shape
+
+        max_mm_seq_len = self.model_config["MAX_MM_SEQ_LEN"]
+        batch_dim = 1 if seq_len < max_mm_seq_len else seq_len // max_mm_seq_len  # Find the division factor
+
+        # TODO: add batch dim for looping along it in case of seq_len > max_mm_seq_len
+        xs = ttnn.reshape(xs, (1, batch_dim, seq_len // batch_dim, -1))
+
+        # Fused QKV
+        fused_query_key_value = ttnn.matmul(
+            xs,
+            self.qkv,
+            dtype=ttnn.bfloat16,
+            memory_config=ttnn.DRAM_MEMORY_CONFIG,
+            compute_kernel_config=self.COMPUTE_KERNEL_QKV,
+        )
+
+        fused_query_key_value = tt_all_reduce(
+            fused_query_key_value, self.device_mesh, cluster_axis=1, num_links=2, memory_config=ttnn.DRAM_MEMORY_CONFIG
+        )
+
+        fused_query_key_value = ttnn.reshape(fused_query_key_value, (1, 1, seq_len, -1))
+
+        xs.deallocate(True)
+
+        (
+            query_layer,  # [bsz, n_local_heads, seq_len, head_dim]
+            key_layer,  # [bsz, n_local_kv_heads, seq_len, head_dim]
+            value_layer,  # [bsz, n_local_kv_heads, seq_len, head_dim]
+        ) = ttnn.experimental.nlp_create_qkv_heads(
+            fused_query_key_value,
+            num_heads=self.n_local_heads,
+            num_kv_heads=self.n_local_kv_heads,
+            transpose_k_heads=False,
+            memory_config=ttnn.DRAM_MEMORY_CONFIG,
+        )
+
+        fused_query_key_value.deallocate(True)
+
+        # ROTARY EMBEDDINGS
+        # Q Rotary Embeddings
+        # query_layer: ttnn.Shape([1, 8, seq_len, 128]) -> [bsz, n_local_heads, seq_len, head_dim]
+        query_layer_ret = ttnn.experimental.rotary_embedding_llama(
+            query_layer, rot_mats[0], rot_mats[1], self.transformation_mats
+        )
+        query_layer.deallocate(True)
+
+        # K Rotary Embeddings
+        # key_layer: ttnn.Shape([1, 1, seq_len, 128]) -> [bsz, n_local_kv_heads, seq_len, head_dim]
+        key_layer_ret = ttnn.experimental.rotary_embedding_llama(
+            key_layer, rot_mats[0], rot_mats[1], self.transformation_mats
+        )
+        key_layer.deallocate(True)
+
+        # query_layer = ttnn.matmul(
+        #     query_layer,
+        #     rot_mats,
+        #     # program_config=self.model_config["ROT_MAT_MM_PROGCFG"],
+        #     memory_config=ttnn.DRAM_MEMORY_CONFIG,
+        #     compute_kernel_config=self.COMPUTE_KERNEL_ROTARY,
+        # )
+
+        # key_layer = ttnn.matmul(
+        #     key_layer,
+        #     rot_mats,
+        #     memory_config=ttnn.DRAM_MEMORY_CONFIG,
+        #     compute_kernel_config=self.COMPUTE_KERNEL_ROTARY,
+        # )
+
+        return query_layer_ret, key_layer_ret, value_layer
+
+    def prefill_prepare_tensor_for_kv_cache(self, key_or_value_layer, user_id):
+        tensor_copy = ttnn.clone(key_or_value_layer)
```

Comment:
Why do we need to clone the kvcache? We'd like to update the original one, right? And just get the references to the tensors belonging to the correct user instead of creating new ones?

Reply:
We are cloning since we are getting tensors from it and creating partial multi-device tensor (8x1 mesh) and the very same tensor is later used in sdpa.